File: yabause/src/android/app/src/main/java/org/uoyabause/android/StateItemAdapter.java
package org.uoyabause.android;

import android.content.Context;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.request.RequestOptions;

import org.devmiyax.yabasanshiro.R;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;

public class StateItemAdapter extends RecyclerView.Adapter<StateItemAdapter.ViewHolder> implements View.OnClickListener {
    private static final String TAG = "CustomAdapter";
    private Yabause _yabause;
    private ArrayList<StateItem> _state_items = null;
    private RecyclerView mRecycler;
    private int selectpos=0;
    public void setSelected( int pos ){
        selectpos = pos;
    }

    public void setStateItems( ArrayList<StateItem> state_items ){
        this._state_items = state_items;
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        mRecycler= recyclerView;
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        mRecycler = null;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView textView;
        private final ImageView imageView;
        private final CardView cardView;
        //private final Toolbar toolbar;

        public ViewHolder(View v) {
            super(v);
            v.setClickable(true);
            textView = (TextView) v.findViewById(R.id.textView);
/*
            toolbar = (Toolbar) v.findViewById(R.id.card_toolbar);
            if (toolbar != null) {
                // inflate your menu
                toolbar.inflateMenu(R.menu.state);
                toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem menuItem) {
                        return true;
                    }
                });
            }
*/
            imageView = (ImageView) v.findViewById(R.id.imageView);
            cardView = (CardView) v.findViewById(R.id.cardview);
        }
        public TextView getTextView() {
            return textView;
        }
        public ImageView getImageView() {
            return imageView;
        }
        public CardView getCardView() {
            return cardView;
        }
        //public Toolbar getToolBar() {
        //    return toolbar;
        //}
    }

    private OnItemClickListener mListener;

    public void setOnItemClickListener(OnItemClickListener listener) {
        mListener = listener;
    }

    public static interface OnItemClickListener {
        public void onItemClick(StateItemAdapter adapter, int position, StateItem item);
    }

    @Override
    public void onClick(View view) {
        if (mRecycler == null) {
            return;
        }
        if (mListener != null) {
            int position = mRecycler.getChildAdapterPosition(view);
            StateItem item = _state_items.get(position);
            mListener.onItemClick(this, position, item);
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        // Create a new view.
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.state_item, viewGroup, false);


        v.setOnClickListener(this);
        v.setFocusable(true);
        v.setFocusableInTouchMode(true);
        return new ViewHolder(v);
    }

    public int dp2px(Context cx, int dp) {
        WindowManager wm = (WindowManager) cx
                .getSystemService(Context.WINDOW_SERVICE);
        Display display = wm.getDefaultDisplay();
        DisplayMetrics displaymetrics = new DisplayMetrics();
        display.getMetrics(displaymetrics);
        return (int) (dp * displaymetrics.density + 0.5f);
    }

    static public final String DATE_PATTERN ="yyyy/MM/dd HH:mm:ss";

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, final int position) {
        Log.d(TAG, "Element " + position + " set.");
        viewHolder.getTextView().setText(new SimpleDateFormat(DATE_PATTERN).format(_state_items.get(position)._savedate));
        //viewHolder.getToolBar().setTitle(new SimpleDateFormat(DATE_PATTERN).format(_state_items.get(position)._savedate));
        Context cx = viewHolder.getImageView().getContext();
        Glide.with(cx)
                .load(new File(_state_items.get(position)._image_filename))
                .apply(new RequestOptions().transforms(new CenterCrop() ))
                .into(viewHolder.getImageView());

                        //.override(dp2px(cx,220),dp2px(cx,220)*3/4)

        if( selectpos == position ){
            viewHolder.getCardView().setBackgroundColor( cx.getResources().getColor(R.color.selected_background) );
            viewHolder.getCardView().setSelected(true);
        }else{
            viewHolder.getCardView().setBackgroundColor( cx.getResources().getColor(R.color.default_background) );
            viewHolder.getCardView().setSelected(false);
        }
    }

    @Override
    public int getItemCount() {
        return _state_items.size();
    }

    public void remove( int position ) {
        File file = new File(_state_items.get(position)._filename);
        if( file != null ){
            file.delete();
        }
        file = new File(_state_items.get(position)._image_filename);
        if( file != null ){
            file.delete();
        }
        _state_items.remove(position);
        notifyItemRemoved(position);
    }
}

File: yabause/src/android/app/src/main/java/org/uoyabause/android/StateItemAdapter.kt
package org.uoyabause.android

import android.content.Context
import android.util.DisplayMetrics
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.request.RequestOptions
import org.devmiyax.yabasanshiro.R
import java.io.File
import java.text.SimpleDateFormat
import java.util.ArrayList

class StateItemAdapter : RecyclerView.Adapter<StateItemAdapter.ViewHolder>(), View.OnClickListener {
    private val _yabause: Yabause? = null
    private var _state_items: ArrayList<StateItem>? = null
    private var mRecycler: RecyclerView? = null
    private var selectpos = 0
    fun setSelected(pos: Int) {
        selectpos = pos
    }

    fun setStateItems(state_items: ArrayList<StateItem>?) {
        _state_items = state_items
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        mRecycler = recyclerView
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        mRecycler = null
    }

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        val textView: TextView
        val imageView: ImageView

        //public Toolbar getToolBar() {
        val cardView: CardView

        //    return toolbar;
        //}
        //private final Toolbar toolbar;
        init {
            v.isClickable = true
            textView = v.findViewById<View>(R.id.textView) as TextView
            /*
            toolbar = (Toolbar) v.findViewById(R.id.card_toolbar);
            if (toolbar != null) {
                // inflate your menu
                toolbar.inflateMenu(R.menu.state);
                toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem menuItem) {
                        return true;
                    }
                });
            }
*/imageView = v.findViewById<View>(R.id.imageView) as ImageView
            cardView = v.findViewById<View>(R.id.cardview) as CardView
        }
    }

    private var mListener: OnItemClickListener? = null
    fun setOnItemClickListener(listener: OnItemClickListener?) {
        mListener = listener
    }

    interface OnItemClickListener {
        fun onItemClick(adapter: StateItemAdapter?, position: Int, item: StateItem?)
    }

    override fun onClick(view: View) {
        if (mRecycler == null) {
            return
        }
        if (mListener != null) {
            val position = mRecycler!!.getChildAdapterPosition(view)
            val item = _state_items!![position]
            mListener!!.onItemClick(this, position, item)
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        // Create a new view.
        val v = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.state_item, viewGroup, false)
        v.setOnClickListener(this)
        v.isFocusable = true
        v.isFocusableInTouchMode = true
        return ViewHolder(v)
    }

    fun dp2px(cx: Context, dp: Int): Int {
        val wm = cx
            .getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = wm.defaultDisplay
        val displaymetrics = DisplayMetrics()
        display.getMetrics(displaymetrics)
        return (dp * displaymetrics.density + 0.5f).toInt()
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        Log.d(TAG, "Element $position set.")
        viewHolder.textView.text = SimpleDateFormat(DATE_PATTERN).format(
            _state_items!![position]._savedate)
        //viewHolder.getToolBar().setTitle(new SimpleDateFormat(DATE_PATTERN).format(_state_items.get(position)._savedate));
        val cx = viewHolder.imageView.context
        Glide.with(cx)
            .load(File(_state_items!![position]._image_filename))
            .apply(RequestOptions().transforms(CenterCrop()))
            .into(viewHolder.imageView)

        //.override(dp2px(cx,220),dp2px(cx,220)*3/4)
        if (selectpos == position) {
            viewHolder.cardView.setBackgroundColor(cx.resources.getColor(R.color.selected_background))
            viewHolder.cardView.isSelected = true
        } else {
            viewHolder.cardView.setBackgroundColor(cx.resources.getColor(R.color.default_background))
            viewHolder.cardView.isSelected = false
        }
    }

    override fun getItemCount(): Int {
        return _state_items!!.size
    }

    fun remove(position: Int) {
        var file = File(_state_items!![position]._filename)
        file?.delete()
        file = File(_state_items!![position]._image_filename)
        file?.delete()
        _state_items!!.removeAt(position)
        notifyItemRemoved(position)
    }

    companion object {
        private const val TAG = "CustomAdapter"
        const val DATE_PATTERN = "yyyy/MM/dd HH:mm:ss"
    }
}